Handle multiple sub prompts for network_cli connection

ganeshrn opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Fixes #35349

*  Add prompt_retry_check to indicate if matched prompt needs to be checked 
    second time to handle invalid answer case.
*  Make required changes in ios and eos terminal plugin to handle
   wrong enable password correctly and return proper error
   message to user.

*  Check if the same prompt is repeated in consecutive window
   if it is repeated it indicates there is the problem with an answer
   provided

*  In that case report error to user
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
network_cli
terminal/ios
terminal/eos
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
Before
```
TASK [Config Using Core Module] ************************************************************************************************************************************************************
task path: /Users/gnalawad/targets/ios_interface.yml:21
<ios> attempting to start connection
<ios> using connection plugin network_cli
<ios> found existing local domain socket, using it!
<ios>
The full traceback is:
Traceback (most recent call last):
  File "/Users/home/ansible/bin/ansible-connection", line 253, in main
    messages.extend(conn.update_play_context(pc_data))
  File "/Users/home/ansible/lib/ansible/module_utils/connection.py", line 146, in __rpc__
    raise ConnectionError(to_text(msg, errors='surrogate_then_replace'), code=code)
ConnectionError: 'NoneType' object has no attribute 'endswith'

fatal: [ios]: FAILED! => {
    "failed": true,
    "msg": "'NoneType' object has no attribute 'endswith'"
```

After:
```
task path: /Users/home/targets/ios_interface.yml:21
<ios> attempting to start connection
<ios> using connection plugin network_cli
<ios> found existing local domain socket, using it!
<ios>
The full traceback is:
Traceback (most recent call last):
  File "/Users/home/ansible/bin/ansible-connection", line 253, in main
    messages.extend(conn.update_play_context(pc_data))
  File "/Users/home/ansible/lib/ansible/module_utils/connection.py", line 146, in __rpc__
    raise ConnectionError(to_text(msg, errors='surrogate_then_replace'), code=code)
ConnectionError: unable to elevate privilege to enable mode, at prompt [
ios>] with error: For matched prompt '
Password: ', answer is not valid

fatal: [ios]: FAILED! => {
    "failed": true,
    "msg": "unable to elevate privilege to enable mode, at prompt [\n38.145.34.223>] with error: For matched prompt '\nPassword: ', answer is not valid"
}
```

ganeshrn committed:
Handle multiple sub prompts for network_cli connection

Fixes #35349

*  Check if the same prompt is repeated in consecutive window
   if it is repeated it indicates there is problem with answer
   provided
*  In that case report error to user

ganeshrn commented:
@bdowling Can you please check if the issue you are facing in https://github.com/ansible/ansible/issues/34719 is fixed by this PR

bdowling commented:
@ganeshrn -- This does address this particular issue, but I do have some concerns.

I think we agree that the real fix for the password prompt should not be to retry, but to gracefully fail if it is seen a second time.  It would be better if on the second occurrence you could immediately error out and inform the user of incorrect password.  Otherwise you risk the user locking out their accounts if they have retry protections in place.  I know this is not what my other PR did either, but I was taking the shortest path to merge. :)

Additionally, I have noted that the handling of prompts passed into receive has a disconnect between the responses, and now it has a added separation between repeats and newline that was recently added.  This prevents you from expecting/handling multiple prompts with different responses for each.  Yes, the occurrence of this situation may be rare, but the additional special cases being added to receive as "global" flags to address particular corner cases I think shows this should maybe be readdressed.

In my opinion, if this is to be added, prompts should be reworked to be a list of dictionaries with each dict providing their prompt, response and other attributes self-contained, and also expose these attributes to the user in playbooks.

If I were to take this a step further, as mentioned to handle the error condition, I could see a couple ways to handle this.  Some other flag could be added fail on a second occurrence (simple, but short sighted), another option is that the list of prompts could optionally be a list of lists of dict, with each grouping having their own handled tracking.  e.g. if group1 was "handled" group2 would be looked for next and you could have rather than a prompt, a new match and custom error message to pass back similar to `terminal_stderr_re` that triggers the failure condition.

ganeshrn commented:
@bdowling Agreed on locking out the user on sending the wrong password multiple time. Currently, the prompt is checked if it repeated in every window received after the first prompt is matched, this might result in the wrong match in the subsequent window down the response.

To avoid this I have added a more strict check to ensure after the first prompt match second match is checked only in the consecutive window. If the prompt is repeated in two consecutive windows it is safe to assume that there is a problem with the answer provided by the user in the input task and should be reported back.

ganeshrn committed:
Fix CI failure

ganeshrn commented:
@bdowling Changing the prompt structure will end breaking backward compatibility as the expected structure of `prompt` from the task is either string or list of strings.

ganeshrn committed:
Fixes #35349

*  Add prompt_retry count to control max number of times
   to expect the same prompt before it error's out

*  Make required changes in ios and eos terminal plugin to handle
   wrong enable password correctly and return proper error
   message to user.

*  Check if the same prompt is repeated in consecutive window
   if it is repeated it indicates there is the problem with an answer
   provided

*  In that case report error to user

ganeshrn commented:
Fix in PR #34719 is breaking ios_user integration test.

Re-running  it with this patch resolves the issue.
```
$ ansible-playbook  -i  ~/targets/inventory.network ../ios.yaml  -e "limit_to=ios_user" -vvvv
<--snip-->
PLAY RECAP ***********************************************************************************************************************************************
ios              : ok=39   changed=11   unreachable=0    failed=0
```

bdowling commented:
The thought occurred to me this morning, why not add the regex for two password prompts in a row into the stderr re list?  This does work, but I needed to change the order of looking for `find_prompt` and `_handle_prompt` so that it was effective.  Also, I am not sure I understand why 'errored_response' is only set after seeing the next CLI prompt, this also caused an issue with this test.  If an stderr re errored response was seen at all, shouldn't things just stop there?

```diff
+++ b/lib/ansible/plugins/terminal/ios.py
@@ -38,6 +38,7 @@ class TerminalModule(TerminalBase):
         # re.compile(br"^% \w+", re.M),
         re.compile(br"% ?Bad secret"),
         re.compile(br"[\r\n%] Bad passwords"),
+        re.compile(br"Password: [\r\n]+Password: ?$"),
         re.compile(br"invalid input", re.I),
         re.compile(br"(?:incomplete|ambiguous) command", re.I),
         re.compile(br"connection timed out", re.I),
@@ -74,7 +75,7 @@ class TerminalModule(TerminalBase):
             if prompt is None or not prompt.endswith(b'#'):
                 raise AnsibleConnectionFailure('failed to elevate privilege to enable mode still at prompt [%s]' % prompt)
         except AnsibleConnectionFailure as e:
-            prompt = self._get_prompt()
+            prompt = self._connection._matched_prompt
             raise AnsibleConnectionFailure('unable to elevate privilege to enable mode, at prompt [%s] with error %s' % (prompt, e.message))

     def on_unbecome(self):
```

```diff
@@ -486,14 +487,15 @@ class Connection(ConnectionBase):
         for regex in self._terminal.terminal_stderr_re:
             if regex.search(response):
                 is_error_message = True
+                display.debug("CLI-PROMPT: Matched Error Prompt %s" % regex.pattern)

                 # Check if error response ends with command prompt if not
                 # receive it buffered prompt
                 for regex in self._terminal.terminal_stdout_re:
                     match = regex.search(response)
+                    errored_response = response
                     if match:
-                        display.debug("CLI-PROMPT: Matched Error Prompt %s" % regex.pattern)
-                        errored_response = response
+                        display.debug("CLI-PROMPT: (after error) Matched Standard Prompt '%s'" % regex.pattern)
                         self._matched_prompt = match.group()
                         break
```

ganeshrn commented:
imo changing the order of `_handle_prompt` and `_find_prompt` is not the right approach. After executing a command the first thing to look for is, command prompt if it is expected (`_handle_prompt` ) and after that cli prompt (`_find_prompt`).

bdowling commented:
The only reason for that order change in my test was because you want it to check for errors.  Perhaps checking stderr re should be a separate function and that is called first.

re: changing the prompt structure.  Both forms could be supported for some time for backwards compatibility.

ganeshrn commented:
I am checking if the command prompt is repeated due to bad answer which is different from checking for error CLI prompt. Adding double command prompt regex for `Password` in `terminal_stderr_re` is not generic to solve the issue of a bad answer (which can be given as input in task).
